convex_hull of polytopes

The documentation mentions that convex_hull is used to compute the convex hull of a polygon (2D only). However, convex_hull appears to run on a VPolytope of any dimensionality. The following quick check shows that adding 2 more dimensions to a vector of 2D points does not change the convex hull, suggesting that the function is not applicable at >2D.
```
V1 = VPolytope([rand(4) for i in 1:200])
V2 = VPolytope([v[1:2] for v in V1.vertices])

length(convex_hull(V1.vertices)) == length(convex_hull(V2.vertices)) #returns true
```
If this is indeed the case, then perhaps convex_hull should contain an assertion for dimensionality to prevent erroneous usage.

Also, given that convex_hull should not be used on higher dimensional polytopes, is there another way to calculate a convex hull? I am aware of the ConvexHull and ConvexHullArray types, but they seem to only work lazily. Is there a way to concretize a ConvexHull into an actual array of points?

Thanks!

To summarize the changes in `convex_hull` for points and for polytopes,  after @tomerarnon 's [first comment](https://github.com/JuliaReach/LazySets.jl/issues/941#issue-391655038) in this thread,  (`master` branch, not tagged yet), let me illustrate with an example in the REPL:

```julia
julia> using LazySets

julia> V1 = VPolytope([rand(4) for i in 1:200]);

julia> V2 = VPolytope([v[1:2] for v in V1.vertices]);

julia> length(convex_hull(V1.vertices)) == length(convex_hull(V2.vertices))  # new
ERROR: AssertionError: this function can only be used for 2-dimensional point clouds; try converting the points to a `VPolytope` first
Stacktrace:
 [1] #convex_hull!#77(::String, ::Function, ::Array{Array{Float64,1},1}) at /Users/forets/.julia/dev/LazySets/src/convex_hull_algorithms.jl:77
 [2] #convex_hull#76 at ./none:0 [inlined]
 [3] convex_hull(::Array{Array{Float64,1},1}) at /Users/forets/.julia/dev/LazySets/src/convex_hull_algorithms.jl:45
 [4] top-level scope at none:0

julia> convex_hull(V1, V2) # new, reminds that Polyhedra dependency is needed
ERROR: AssertionError: this function needs the package 'Polyhedra' to be loaded
Stacktrace:
 [1] default_polyhedra_backend(::VPolytope{Float64}, ::Type) at /Users/forets/.julia/dev/LazySets/src/AbstractPolytope.jl:111
 [2] convex_hull(::VPolytope{Float64}, ::VPolytope{Float64}) at /Users/forets/.julia/dev/LazySets/src/VPolytope.jl:216
 [3] top-level scope at none:0

julia> using Polyhedra, CDDLib   # CDDLib is recommended for convex_hull, for performance

julia> Vch = convex_hull(V1, V2, backend=CDDLib.Library()); # first time slow, then fast

julia> length(Vch.vertices)
81
```

--- a/docs/src/lib/binary_functions.md
+++ b/docs/src/lib/binary_functions.md
@@ -53,7 +53,7 @@ convex_hull(::HPoly{N}, ::HPoly{N}) where {N<:Real}
 convex_hull(::VPolytope{N}, ::VPolytope{N}) where {N<:Real}
 convex_hull(::VPolygon{N}, ::VPolygon{N}) where {N<:Real}
 convex_hull(::Vector{VN}) where {N<:Real, VN<:AbstractVector{N}}
-convex_hull(::UnionSetArray{N, PT}) where {N, PT<:AbstractPolytope{N}}
+convex_hull(::UnionSetArray{N, PT}; kwargs...) where {N, PT<:AbstractPolytope{N}}
 monotone_chain!
 ```
 

--- a/docs/src/lib/binary_functions.md
+++ b/docs/src/lib/binary_functions.md
@@ -53,7 +53,7 @@ convex_hull(::HPoly{N}, ::HPoly{N}) where {N<:Real}
 convex_hull(::VPolytope{N}, ::VPolytope{N}) where {N<:Real}
 convex_hull(::VPolygon{N}, ::VPolygon{N}) where {N<:Real}
 convex_hull(::Vector{VN}) where {N<:Real, VN<:AbstractVector{N}}
-convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}; kwargs...) where {N}
+convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}) where {N}
 monotone_chain!
 ```
 

--- a/docs/src/lib/binary_functions.md
+++ b/docs/src/lib/binary_functions.md
@@ -53,7 +53,7 @@ convex_hull(::HPoly{N}, ::HPoly{N}) where {N<:Real}
 convex_hull(::VPolytope{N}, ::VPolytope{N}) where {N<:Real}
 convex_hull(::VPolygon{N}, ::VPolygon{N}) where {N<:Real}
 convex_hull(::Vector{VN}) where {N<:Real, VN<:AbstractVector{N}}
-convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}) where {N}
+convex_hull(U::UnionSetArray{N, PT}) where {N, PT<:AbstractPolytope{N}}
 monotone_chain!
 ```
 

--- a/docs/src/lib/binary_functions.md
+++ b/docs/src/lib/binary_functions.md
@@ -53,6 +53,7 @@ convex_hull(::HPoly{N}, ::HPoly{N}) where {N<:Real}
 convex_hull(::VPolytope{N}, ::VPolytope{N}) where {N<:Real}
 convex_hull(::VPolygon{N}, ::VPolygon{N}) where {N<:Real}
 convex_hull(::Vector{VN}) where {N<:Real, VN<:AbstractVector{N}}
+convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}; kwargs...) where {N}
 monotone_chain!
 ```
 

--- a/docs/src/lib/binary_functions.md
+++ b/docs/src/lib/binary_functions.md
@@ -53,7 +53,7 @@ convex_hull(::HPoly{N}, ::HPoly{N}) where {N<:Real}
 convex_hull(::VPolytope{N}, ::VPolytope{N}) where {N<:Real}
 convex_hull(::VPolygon{N}, ::VPolygon{N}) where {N<:Real}
 convex_hull(::Vector{VN}) where {N<:Real, VN<:AbstractVector{N}}
-convex_hull(U::UnionSetArray{N, PT}) where {N, PT<:AbstractPolytope{N}}
+convex_hull(::UnionSetArray{N, PT}) where {N, PT<:AbstractPolytope{N}}
 monotone_chain!
 ```
 

--- a/src/ConcreteOperations/convex_hull.jl
+++ b/src/ConcreteOperations/convex_hull.jl
@@ -478,8 +478,8 @@ function monotone_chain!(points::Vector{VN}; sort::Bool=true
 end
 
 """
-    convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}) where {N}
-    
+    convex_hull(U::UnionSetArray{N, PT}) where {N, PT<:AbstractPolytope{N}}
+
 Compute the convex hull of a union of a finite number of polytopes.
 
 ### Input
@@ -490,7 +490,7 @@ Compute the convex hull of a union of a finite number of polytopes.
 
 A list of the vertices of the convex hull.
 """
-function convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}) where {N}
+function convex_hull(U::UnionSetArray{N, PT}) where {N, PT<:AbstractPolytope{N}}
     vlist = mapreduce(vertices_list, vcat, U.array)
     return convex_hull(vlist)
 end

--- a/src/ConcreteOperations/convex_hull.jl
+++ b/src/ConcreteOperations/convex_hull.jl
@@ -476,3 +476,21 @@ function monotone_chain!(points::Vector{VN}; sort::Bool=true
     end
     return resize!(points, m)
 end
+
+"""
+    convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}; kwargs...) where {N}
+Compute the convex hull of an array of polytopes
+
+### Input
+
+- `U` -- UnionSetArray of polytopes
+
+### Output
+
+List of vectors containing the vertices of the convex hull of an array of
+the given polytopes.
+"""
+function convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}; kwargs...) where {N}
+    vlist = mapreduce(vertices_list, vcat, U.array)
+    return convex_hull(vlist; kwargs...)
+end

--- a/src/ConcreteOperations/convex_hull.jl
+++ b/src/ConcreteOperations/convex_hull.jl
@@ -478,7 +478,7 @@ function monotone_chain!(points::Vector{VN}; sort::Bool=true
 end
 
 """
-    convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}; kwargs...) where {N}
+    convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}) where {N}
 Compute the convex hull of an array of polytopes
 
 ### Input
@@ -490,7 +490,7 @@ Compute the convex hull of an array of polytopes
 List of vectors containing the vertices of the convex hull of an array of
 the given polytopes.
 """
-function convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}; kwargs...) where {N}
+function convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}) where {N}
     vlist = mapreduce(vertices_list, vcat, U.array)
-    return convex_hull(vlist; kwargs...)
+    return convex_hull(vlist)
 end

--- a/src/ConcreteOperations/convex_hull.jl
+++ b/src/ConcreteOperations/convex_hull.jl
@@ -478,7 +478,7 @@ function monotone_chain!(points::Vector{VN}; sort::Bool=true
 end
 
 """
-    convex_hull(U::UnionSetArray{N, PT}) where {N, PT<:AbstractPolytope{N}}
+    convex_hull(U::UnionSetArray{N, PT}; kwargs...) where {N, PT<:AbstractPolytope{N}}
 
 Compute the convex hull of a union of a finite number of polytopes.
 
@@ -490,7 +490,7 @@ Compute the convex hull of a union of a finite number of polytopes.
 
 A list of the vertices of the convex hull.
 """
-function convex_hull(U::UnionSetArray{N, PT}) where {N, PT<:AbstractPolytope{N}}
+function convex_hull(U::UnionSetArray{N, PT}; kwargs...) where {N, PT<:AbstractPolytope{N}}
     vlist = mapreduce(vertices_list, vcat, U.array)
-    return convex_hull(vlist)
+    return convex_hull(vlist; kwargs...)
 end

--- a/src/ConcreteOperations/convex_hull.jl
+++ b/src/ConcreteOperations/convex_hull.jl
@@ -479,7 +479,8 @@ end
 
 """
     convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}) where {N}
-Compute the convex hull of an array of polytopes
+    
+Compute the convex hull of a union of a finite number of polytopes.
 
 ### Input
 
@@ -487,8 +488,7 @@ Compute the convex hull of an array of polytopes
 
 ### Output
 
-List of vectors containing the vertices of the convex hull of an array of
-the given polytopes.
+A list of the vertices of the convex hull.
 """
 function convex_hull(U::UnionSetArray{N, <:AbstractPolytope{N}}) where {N}
     vlist = mapreduce(vertices_list, vcat, U.array)

--- a/test/unit_convex_hull.jl
+++ b/test/unit_convex_hull.jl
@@ -152,4 +152,10 @@ for N in [Float64, Rational{Int}]
     S = Singleton(N[1])
     ch = convex_hull(S, S)
     @test ch isa Singleton{N} && element(ch) == element(S)
+
+    # UnionSetArray
+    V1 = VPolytope([N[0, 0], N[1, 0], N[0, 1]])
+    V2 = VPolytope([N[1, 1], N[1, 0], N[0, 1]])
+    U = UnionSetArray([V1, V2]);
+    @test ispermutation(convex_hull(U), [N[0, 0], N[1, 0], N[0, 1], N[1, 1]])
 end

